### PR TITLE
feat: Notify parties of Order made on C21 (FPLA-603)

### PR DESCRIPTION
**Pull request by Gogetter**
### Notify parties of Order made on C21 ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java [hunk 1]
```diff
@@ -68,43 +84,39 @@ void shouldBuildNotificationTemplatesWithCompleteValues() throws Exception {
...
-            .put("fullStop", "No")
-            .put("ordersAndDirections", ordersAndDirections)
-            .put("timeFramePresent", "Yes")
-            .put("timeFrameValue", "Same day")
-            .put("reference", "12345")
-            .put("caseUrl", "http://fake-url/case/" + JURISDICTION + "/" + CASE_TYPE + "/12345")
+            .put(ORDERS_AND_DIRECTIONS, ordersAndDirections)
             .build();
 
         Map<String, Object> expectedCafcassParameters = ImmutableMap.<String, Object>builder()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: Move to sepearate private method perhaps

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java [hunk 2]
```diff
@@ -68,43 +84,39 @@ void shouldBuildNotificationTemplatesWithCompleteValues() throws Exception {
         List<String> ordersAndDirections = ImmutableList.of("Emergency protection order",
             "Contact with any named person");
         Map<String, Object> expectedHmctsParameters = ImmutableMap.<String, Object>builder()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: Move to sepearate private method perhaps

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java [hunk 3]
```diff
@@ -36,8 +36,24 @@
...
+    private static final String ORDERS_AND_DIRECTIONS = "ordersAndDirections";
+    private static final String CAFCASS = "cafcass";
+    private static final String SUBMITTED_CALLBACK_URL = "/callback/case-submission/submitted";
+    private static final String AUTHORIZATION = "authorization";
+    private static final String USER_ID_STR = "user-id";
+
     private static final ObjectMapper MAPPER = new ObjectMapper();
 
+    private static final Map<String, Object> expectedSimilarParameters = ImmutableMap.<String, Object>builder()
+        .put("localAuthority", "Example Local Authority")
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: Should we call this expectedParameters instead? We don't want similar we want the exact same?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java [hunk 4]
```diff
@@ -25,8 +25,8 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 import static uk.gov.hmcts.reform.fpl.CaseDefinitionConstants.CASE_TYPE;
 import static uk.gov.hmcts.reform.fpl.CaseDefinitionConstants.JURISDICTION;
```
Comment (swalker125) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: why were there changes to this file? Nothing to do with notify c21.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java [hunk 5]
```diff
@@ -25,7 +25,7 @@
 import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
```
Comment (swalker125) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: why were there changes to this file?

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java [hunk 6]
```diff
@@ -123,39 +135,41 @@ void shouldBuildNotificationTemplatesWithValuesMissingInCallback() throws Except
...
             .put("fullStop", "Yes")
-            .put("ordersAndDirections", "")
+            .put(ORDERS_AND_DIRECTIONS, "")
             .put("timeFramePresent", "No")
             .put("timeFrameValue", "")
             .put("reference", "12345")
             .put("caseUrl", "http://fake-url/case/" + JURISDICTION + "/" + CASE_TYPE + "/12345")
             .build();
 
         Map<String, Object> expectedCafcassParameters = ImmutableMap.<String, Object>builder()
```
Comment (toireasam) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: we build these parameters in a couple of methods so private method would be better for this i think

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/NotificationTemplateType.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.hmcts.reform.fpl.enums;
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/NotificationTemplateType.java]: why?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java [hunk 8]
```diff
@@ -128,12 +155,27 @@ public void notifyLocalAuthorityOfIssuedStandardDirectionsOrder(StandardDirectio
...
+                                 final String reference) {
         log.debug("Sending submission notification (with template id: {}) to {}", templateId, email);
         try {
             notificationClient.sendEmail(templateId, email, parameters, reference);
         } catch (NotificationClientException e) {
             log.error("Failed to send submission notification (with template id: {}) to {}", templateId, email, e);
         }
     }
+
+    private void sendNotifications(final String templateId,
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: why not just use .forEach on the list of emails to send notifications and use the existing sendNotifications?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/AbstractEmailContentProvider.java [hunk 9]
```diff
@@ -24,8 +24,8 @@
 public abstract class AbstractEmailContentProvider {
 
     final String uiBaseUrl;
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/AbstractEmailContentProvider.java]: why changes to this file?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+        final String subjectLine = buildSubjectLine(caseData);
+
+        return ImmutableMap.of(
+            "subjectLine", subjectLine,
+            "localAuthorityOrCafcass", localAuthorityNameLookupConfiguration.getLocalAuthorityName(localAuthorityCode),
+            "hearingDetailsCallout", buildSubjectLineWithHearingBookingDateSuffix(subjectLine, caseData),
+            "reference", String.valueOf(caseDetails.getId()),
+            "caseUrl", uiBaseUrl + "/case/" + JURISDICTION + "/" + CASE_TYPE + "/" + caseDetails.getId()
+
```
Comment (toireasam) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: remove whitespace

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C2UploadedEmailContentProvider.java [hunk 11]
```diff
@@ -18,7 +18,7 @@
 @Service
 public class C2UploadedEmailContentProvider extends AbstractEmailContentProvider {
 
-    private final ObjectMapper objectMapper;
+    final ObjectMapper objectMapper;
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C2UploadedEmailContentProvider.java]: why changes to this file?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 12]
```diff
@@ -145,7 +159,38 @@ void shouldNotifyNonHmctsAdminOnC2Upload() throws IOException, NotificationClien
...
+                .willReturn(new Court(COURT_NAME, COURT_EMAIL_ADDRESS));
+
+            given(cafcassLookupConfiguration.getCafcass(LOCAL_AUTHORITY_CODE))
+                .willReturn(new Cafcass(CAFCASS_NAME, CAFCASS_EMAIL_ADDRESS));
+
+            given(localAuthorityNameLookupConfiguration.getLocalAuthorityName(LOCAL_AUTHORITY_CODE))
+                .willReturn("Example Local Authority");
+
+            given(c21OrderEmailContentProvider.buildC21OrderNotification(callbackRequest().getCaseDetails(),
+                LOCAL_AUTHORITY_CODE)).willReturn(parameters);
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: can all of this be moved outside of the test. Lots of clutter. I only care that an email was sent to a party in this test

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java (hunk 1 above)
Comment (Gogetter) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/CaseSubmissionControllerSubmittedTest.java]: its was moved into a separate variable `expectedSimilarParameters`

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java (hunk 5 above)
Comment (Gogetter) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NotifyGatekeeperTest.java]: it was updated following refactoring of `NotificationTemplates` in `NotificationTemplateType` an enum

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java (hunk 8 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: we could actually...

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/AbstractEmailContentProvider.java (hunk 9 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/AbstractEmailContentProvider.java]: in order to reuse instead of doing a `super.uiBaseUrl` when building personalisations for the separate email templates as we do not need all values from `getCasePersonalisationBuilder(...)`

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C2UploadedEmailContentProvider.java (hunk 11 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C2UploadedEmailContentProvider.java]: good spot. This did not need changing

**Comment by Gogetter** [comment]
@swalker125 just made and pushed the changes.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java [hunk 13]
```diff
@@ -82,6 +88,23 @@ public void sendNotificationForC2Upload(final C2UploadedEvent caseEvent) {
...
+
+        Map<String, Object> parameters = c21OrderEmailContentProvider.buildC21OrderNotification(
+            caseDetails, localAuthorityCode);
+        String reference = Long.toString(caseDetails.getId());
+
+        String localAuthorityEmail = hmctsCourtLookupConfiguration.getCourt(localAuthorityCode).getEmail();
+        String cafcassEmail = cafcassLookupConfiguration.getCafcass(localAuthorityCode).getEmail();
+
+        Stream.of(localAuthorityEmail, cafcassEmail)
+            .filter(StringUtils::isNotBlank)
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: will either of these emails ever be null? I don't think so...

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 14]
```diff
@@ -147,6 +176,22 @@ void shouldNotifyNonHmctsAdminOnC2Upload() throws IOException, NotificationClien
             verify(notificationClient, times(1)).sendEmail(
                 eq(C2_UPLOAD_NOTIFICATION_TEMPLATE), eq("hmcts-non-admin@test.com"), eq(parameters), eq("12345"));
         }
+
+        @Test
+        void shouldNotifyPartiesOnC21OrderSubmission() throws IOException, NotificationClientException {
+            given(c21OrderEmailContentProvider.buildC21OrderNotification(callbackRequest().getCaseDetails(),
+                LOCAL_AUTHORITY_CODE)).willReturn(parameters);
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: this could probably be moved as well

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java [hunk 15]
```diff
@@ -180,7 +225,8 @@ void shouldSendEmailToHmcts() throws IOException, NotificationClientException {
         notificationHandler.sendNotificationToHmctsAdmin(new SubmittedCaseEvent(callbackRequest(), AUTH_TOKEN, USER_ID));
 
         verify(notificationClient, times(1)).sendEmail(
-            eq(HMCTS_COURT_SUBMISSION_TEMPLATE), eq(COURT_EMAIL_ADDRESS), eq(expectedParameters), eq("12345"));
+            eq(HMCTS_COURT_SUBMISSION_TEMPLATE), eq(COURT_EMAIL_ADDRESS),
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: there are some mockito stubs above that can be removed now

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java (hunk 13 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: I agree it should not. But no harm checking was my thinking.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java (hunk 15 above)
Comment (Gogetter) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: updated

File: service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java (hunk 14 above)
Comment (Gogetter) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandlerTest.java]: updated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/NotificationTemplateType.java (hunk 7 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/NotificationTemplateType.java]: class now deleted following our conversation

**Comment by Gogetter** [comment]
@swalker125 more updates made and pushed. please could you help give another look? thanks

File: ccd-definition/CaseEvent/CaseEvent.json [hunk 16]
```diff
@@ -511,7 +511,8 @@
     "ShowEventNotes": "N",
     "EndButtonLabel": "Save and continue",
     "CallBackURLAboutToStartEvent": "${CCD_DEF_CASE_SERVICE_BASE_URL}/callback/create-order/about-to-start",
-    "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CASE_SERVICE_BASE_URL}/callback/create-order/about-to-submit"
+    "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CASE_SERVICE_BASE_URL}/callback/create-order/about-to-submit",
+    "CallBackURLSubmittedEvent": "${CCD_DEF_CASE_SERVICE_BASE_URL}/callback/create-order/about-to-submit"
```
Comment (lewisbirks) [inline ccd-definition/CaseEvent/CaseEvent.json]: Change the url to submitted

File: ccd-definition/CaseEvent/CaseEvent.json [hunk 17]
```diff
@@ -705,7 +706,8 @@
...
```
Comment (lewisbirks) [inline ccd-definition/CaseEvent/CaseEvent.json]: Change url to submitted

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java [hunk 18]
```diff
@@ -128,7 +151,8 @@ public void notifyLocalAuthorityOfIssuedStandardDirectionsOrder(StandardDirectio
         sendNotification(STANDARD_DIRECTION_ORDER_ISSUED_TEMPLATE, email, parameters, reference);
     }
 
-    private void sendNotification(String templateId, String email, Map<String, Object> parameters, String reference) {
+    private void sendNotification(final String templateId, final String email, final Map<String, Object> parameters,
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: unnecessary change

File: ccd-definition/CaseEvent/CaseEvent.json (hunk 16 above)
Comment (Gogetter) [reply ccd-definition/CaseEvent/CaseEvent.json]: thanks for spotting. This must have been whilst resolving the merge conflict

File: ccd-definition/CaseEvent/CaseEvent.json (hunk 17 above)
Comment (Gogetter) [reply ccd-definition/CaseEvent/CaseEvent.json]: updated

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java (hunk 13 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: I think the local authority email MAY be null. We have got some follow up stories for SDO, should the same behaviour be applied here? I would say so, please check with BAs

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+                                               localAuthorityNameLookupConfiguration,
+                                           DateFormatterService dateFormatterService) {
+        super(uiBaseUrl, dateFormatterService, hearingBookingService);
+        this.objectMapper = objectMapper;
+        this.localAuthorityNameLookupConfiguration = localAuthorityNameLookupConfiguration;
+    }
+
+    public Map<String, Object> buildC21OrderNotification(final CaseDetails caseDetails,
+                                                         final String localAuthorityCode) {
+        // Validation within our frontend ensures that the following data is present
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: This comment is too generic, maybe would be better to move it above the line to which it applies?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java (hunk 13 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/NotificationHandler.java]: Hmmmm... was not aware of this @akomisarek, will speak to a BA to clarify how we proceed.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java (hunk 19 above)
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: removed in [pr-618](https://github.com/hmcts/fpl-ccd-configuration/pull/618) as its a follow-up to this. Would you want to done here also?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java [hunk 20]
```diff
@@ -25,31 +26,37 @@
...
-            "localAuthorityOrCafcass", localAuthorityNameLookupConfiguration.getLocalAuthorityName(localAuthorityCode),
-            "hearingDetailsCallout", buildSubjectLineWithHearingBookingDateSuffix(subjectLine, caseData),
-            "reference", String.valueOf(caseDetails.getId()),
-            "caseUrl", uiBaseUrl + "/case/" + JURISDICTION + "/" + CASE_TYPE + "/" + caseDetails.getId()
-        );
+    public Map<String, Object> buildC21OrderNotificationParametersForCafcass(final CaseDetails caseDetails,
+                                                                             final String localAuthorityCode) {
+        return ImmutableMap.<String, Object>builder()
+            .putAll(commonC21NotificationParameters(caseDetails))
+            .put("localAuthorityOrCafcass", cafcassLookupConfiguration.getCafcass(localAuthorityCode).getName())
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: That will produce "Cafcass Cymru" are we 100% sure this is what business wants? Can's see this in the AC.

Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: Hi @akomisarek, we have confirmed this is the case from Jay and Mia.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java (hunk 19 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/email/content/C21OrderEmailContentProvider.java]: No it's fine only in PR-618
